pkg/cache/sql/store/store.go

```diff
@@ -0,0 +1,370 @@
+/*
+Package store contains the sql backed store. It persists objects to a sqlite database.
+*/
+package store
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"encoding/gob"
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/lasso/pkg/cache/sql/db"
+	"github.com/rancher/lasso/pkg/cache/sql/db/transaction"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	_ "modernc.org/sqlite"
+)
+
+const (
+	upsertStmtFmt   = `REPLACE INTO "%s"(key, object, objectnonce, dek, deknonce) VALUES (?, ?, ?, ?, ?)`
+	deleteStmtFmt   = `DELETE FROM "%s" WHERE key = ?`
+	getStmtFmt      = `SELECT object, objectnonce, dek, deknonce FROM "%s" WHERE key = ?`
+	listStmtFmt     = `SELECT object, objectnonce, dek, deknonce FROM "%s"`
+	listKeysStmtFmt = `SELECT key FROM "%s"`
+)
+
+// Store is a SQLite-backed cache.Store
+type Store struct {
+	c       cache.Indexer
+	name    string
+	encrypt bool
+	typ     reflect.Type
+	keyFunc cache.KeyFunc
+	dbLock  *sync.RWMutex
+
+	DBClient
+
+	shouldEncrypt bool
+
+	upsertStmt   *sql.Stmt
+	deleteStmt   *sql.Stmt
+	getStmt      *sql.Stmt
+	listStmt     *sql.Stmt
+	listKeysStmt *sql.Stmt
+
+	afterUpsert []func(key string, obj any, tx db.TXClient) error
+	afterDelete []func(key string, tx db.TXClient) error
+}
+
+// Test that Store implements cache.Indexer
+var _ cache.Store = (*Store)(nil)
+
+type DBClient interface {
+	Begin() (db.TXClient, error)
+	Prepare(stmt string) *sql.Stmt
+	QueryForRows(ctx context.Context, stmt transaction.Stmt, params ...any) (*sql.Rows, error)
+	ReadObjects(rows db.Rows, typ reflect.Type, shouldDecrypt bool) ([]any, error)
+	ReadStrings(rows db.Rows) ([]string, error)
+	Upsert(tx db.TXClient, stmt *sql.Stmt, key string, obj any, shouldEncrypt bool) error
```

Comment:
IIUC intent is to abstract away something that can store and retrieve objects SQL storage + blob encoding (gob) + optional encryption "layer"

Encryption ends up determining a part of the structure of tables/columns (`object, dek, deknonce, datanonce`) which ends up influencing upsert, select and create table statements.

So DBClient ends up with knowledge about, and DBClient users have to rely on this implicit assumption.

I will think about if and how this could be decoupled better.

Reply:
I think there are plenty of different ways the db layer could be designed. To be honest the largest value I was looking to extract was testability, so I needed some  of layer there. It provides a database client, encrypting/decrypting like @moio pointed out, some automatic closing, and a connection reset that also cleans files. I think if desired it could always be changed later. I think a challenge in having a different client for each type will be using the same same connection for all of them. In most cases it's not recommended to have more than 1 or 2 (if you want to separate read and write) since a connection in golang's sql package is a pool of connections that are being coordinated for you.

Reply:
-> [epic](https://github.com/rancher/rancher/issues/45635)